### PR TITLE
Fix issue 589

### PR DESCRIPTION
**Pull request by smarie**
only the first item in the following list actually applies :)

# Pull Request Check List

- [x ] Added **tests** for changed code.
- [x ] New features have been added to our [Hypothesis testing strategy](https://github.com/python-attrs/attrs/blob/master/tests/strategies.py).
- [x ] Changes or additions to public APIs are reflected in our type stubs (files ending in ``.pyi``).
    - [x ] ...and used in the stub test file `tests/typing_example.py`.
- [x ] Updated **documentation** for changed code.
    - [x ] New functions/classes have to be added to `docs/api.rst` by hand.
    - [x ] Changes to the signature of `@attr.s()` have to be added by hand too.
    - [x ] Changed/added classes/methods/functions have appropriate `versionadded`, `versionchanged`, or `deprecated` [directives](http://www.sphinx-doc.org/en/stable/markup/para.html#directive-versionadded).
- [x ] Documentation in `.rst` files is written using [semantic newlines](https://rhodesmill.org/brandon/2012/one-sentence-per-line/).
- [x ] Changes (and possible deprecations) have news fragments in [`changelog.d`](https://github.com/python-attrs/attrs/blob/master/changelog.d).

If you have *any* questions to *any* of the points above, just **submit and ask**!  This checklist is here to *help* you, not to deter you from contributing!

### CODE REVIEW

**Comment by smarie** [comment]
hum it seems that some very old mods leaked in the rst file. It should not be part of the PR. I'm trying to remove it right now

**Comment by smarie** [comment]
rather closing and re-creating a proper one: the tree seems quite full of dirt here, sorry.

**Comment by smarie** [comment]
The diff seems ok now. 

Sorry for the bad tree (commits from 2017 mentioned above), if that is a problem I can delete my attrs fork and start from a fresh one.

**Comment by hynek** [comment]
- Could you please add a changelog entry? You've checked it in the checklist but I can't find it. ;)
- Please run Black (or just `tox -e lint`), it's failing the CI.

Thanks!

**Comment by smarie** [comment]
Oh sorry - sure, I'll do.

Actually I'll close this PR and open another one with cleaner history.

**Comment by hynek** [comment]
Don't worry about the history, we squash on merge anyways.

File: tests/test_slots.py [hunk 1]
```diff
@@ -543,3 +543,18 @@ class C(object):
...
+    """
+
+    @attr.s(slots=True)
+    class C(object):
+        field = attr.ib()
+
+        def f(self, a):
+            super(C, self).__init__()
+
+    C(field=1)
```
Comment (smarie) [inline tests/test_slots.py]:
```suggestion
    C(field=1)

```

**Comment by smarie** [comment]
The tox lint config does not seem to take into account proxy environment variables. I am located behind a corporate proxy so it fails with ` fatal: unable to access 'https://github.com/psf/black/': Failed to connect to github.com port 443: Timed out`

I'll try later or when back home. Alternately if you have the opportunity to see the issue in `test_slots.py`, do not hesitate to propose a review comment

Thanks again hynek

**Comment by hynek** [comment]
Feel free to add an `passenv` to tox.ini

File: src/attr/_make.py [hunk 2]
```diff
@@ -640,8 +640,13 @@ def slots_setstate(self, state):
...
             for cell in closure_cells:
-                if cell.cell_contents is self._cls:
-                    set_closure_cell(cell, cls)
+                try:
+                    match = cell.cell_contents is self._cls
+                except ValueError:  # ValueError: Cell is empty
+                    pass
+                else:
+                    if match:
+                        set_closure_cell(cell, cls)
```
Comment (hynek) [inline src/attr/_make.py]: Can't you just move this into the try block?

File: changelog.d/590.change.rst [hunk 3]
```diff
@@ -0,0 +1 @@
+Fixed a `ValueError: Cell is empty` bug that could happen in some rare edge cases.
```
Comment (hynek) [inline changelog.d/590.change.rst]:
```suggestion
Fixed a ``ValueError: Cell is empty`` bug that could happen in some rare edge cases.
```

[the remark this answers is not in the document]
Comment (smarie) [reply changelog.d/590.change.rst]: I can put double backquotes if that's what you mean. The default implied reST role will be `:literal:` if I read the [doc](https://docutils.sourceforge.io/docs/ref/rst/roles.html#literal) correctly. If you'd like another role just let me know

File: tests/test_slots.py [hunk 4]
```diff
@@ -543,3 +543,18 @@ class C(object):
     w = weakref.ref(c)
 
     assert c is w()
+
+
+def test_slots_empty_cell():
+    """
+    test for issue https://github.com/python-attrs/attrs/issues/589
```
[the remark this answers is not in the document]
Comment (smarie) [reply tests/test_slots.py]: ok

File: changelog.d/590.change.rst (hunk 3 above)
[the remark this answers is not in the document]
Comment (smarie) [reply changelog.d/590.change.rst]: oh I see you did it yourself. thanks

File: src/attr/_make.py (hunk 2 above)
Comment (smarie) [reply src/attr/_make.py]: I'd rather not - if a `ValueError` is raised by `set_closure_cell(...)` one day, this is maybe something you'd like to know asap, not after investigating a potential side-effect that may pop up somewhere else and take longer to be reported as an issue. But of course that's your decision, let me know.

**Comment by smarie** [comment]
Should be fine now. Do not hesitate to change the test description if that does not suit you - I was rather verbose :)

**Comment by hynek** [comment]
Line 559 is too long 🙈

File: tests/test_slots.py [hunk 5]
```diff
@@ -543,3 +543,27 @@ class C(object):
...
+    Tests that no `ValueError: Cell is empty` exception is raised when
+    closure cells are present with no contents in a `slots=True` class.
+    (issue https://github.com/python-attrs/attrs/issues/589)
+
+    On Python 3, if a method mentions `__class__` or uses the no-arg `super()`,
+    the compiler will bake a reference to the class in the method itself as
+    `method.__closure__`. Since `attrs` replaces the class with a clone,
+    `_ClassBuilder._create_slots_class(self)` will rewrite these references so
+    it keeps working. This method was not properly covering the edge case where
+    the closure cell was empty, we fixed it and this is the non-regression test.
```
Comment (smarie) [inline tests/test_slots.py]:
```suggestion
    the closure cell was empty, we fixed it and this is the non-regression 
    test.
```

**Comment by smarie** [comment]
That's what happens when I try to complete a PR while doing other things in parallel :) sorry hynek. This is now ok.

**Comment by hynek** [comment]
Thanks!
